core/src/main/java/hudson/ClassicPluginStrategy.java

```diff
@@ -490,21 +494,48 @@ private static void explode(File archive, File destDir) throws IOException {
         // delete the contents so that old files won't interfere with new files
         Util.deleteRecursive(destDir);
 
+        File destDirPreTransform = new File(destDir.getParentFile(), destDir.getName() + ".pre-transform");
+        Util.createDirectories(Util.fileToPath(destDirPreTransform));
         try {
             Project prj = new Project();
-            unzipExceptClasses(archive, destDir, prj);
-            createClassJarFromWebInfClasses(archive, destDir, prj);
+            unzipExceptClasses(archive, destDirPreTransform, prj);
+            createClassJarFromWebInfClasses(archive, destDirPreTransform, prj);
         } catch (BuildException x) {
             throw new IOException("Failed to expand " + archive,x);
         }
 
+        transform(destDirPreTransform.getAbsolutePath(), destDir.getAbsolutePath());
+
         try {
             new FilePath(explodeTime).touch(archive.lastModified());
         } catch (InterruptedException e) {
             throw new AssertionError(e); // impossible
         }
     }
 
+    private static void transform(String... args) {
+        Transformer jTrans = new Transformer(System.out, System.err);
+        jTrans.setOptionDefaults(ClassicPluginStrategy.class, getOptionDefaults());
+        jTrans.setArgs(args);
+
+        @SuppressWarnings("unused")
+        int rc = jTrans.run();
+        if (rc != 0) {
+            throw new RuntimeException("Transformer failed with exit code: " + rc);
+        }
+    }
+
+    private static Map<AppOption, String> getOptionDefaults() {
+        Map<AppOption, String> optionDefaults = new HashMap<>();
+        optionDefaults.put(AppOption.RULES_RENAMES, "jakarta-renames.properties");
+        optionDefaults.put(AppOption.RULES_VERSIONS, "jakarta-versions.properties");
+        optionDefaults.put(AppOption.RULES_BUNDLES, "jakarta-bundles.properties");
+        optionDefaults.put(AppOption.RULES_DIRECT, "jakarta-direct.properties");
+        optionDefaults.put(AppOption.RULES_MASTER_TEXT, "jakarta-text-master.properties");
+        optionDefaults.put(AppOption.RULES_PER_CLASS_CONSTANT, "jakarta-per-class-constant-master.properties");
+        return Collections.unmodifiableMap(optionDefaults);
+    }
```

Comment:
Beginnings of a compatibility layer for plugins. This initial prototype uses [Eclipse Transformer](https://github.com/eclipse/transformer) to transform the extracted plugin to a Jakarta EE 9 compatible version after exploding the plugin JPI. Some possible ideas for how to productize this:

- Perhaps we might update `plugin-pom`/`maven-hpi-plugin` to apply this transformation when the plugin is built and then mark the plugin as having been transformed in `MANIFEST.MF`. Then we can skip over the transformation for that plugin at runtime.
- Perhaps it's too heavyweight to rewrite all the plugins' contents at startup. The Eclipse Transformer `README` lists a "possible secondary use" of dynamic rewriting at class loading time. Perhaps we want to lazily do the transformation at class loading time rather than eagerly transforming all plugins when Jenkins starts up.

Reply:
> Perhaps we might update `plugin-pom`/`maven-hpi-plugin` to apply this transformation when the plugin is built and then mark the plugin as having been transformed in `MANIFEST.MF`. Then we can skip over the transformation for that plugin at runtime.

Is the Servlet API available to pluginsnot tied to the core version the plugin depends on? If it is, I imagine at some point a plugin maintainer raises the Jenkins core dependency to whichever release includes this change, and then the plugin should no longer build unless imports are fixed. Do you want maintainers to not have to do that? Because declaring a lesser Jenkins core dependency, and then unconditionally rewriting imports would be incompatible with that core.